File: shared/js/frame.mjs
import ruleIds from "../frame-rules.json" with { type: "json" };

export const DEFAULT_MAX_FRAME_BYTES = 32 * 1024 * 1024;
export const MAX_HEADER_BYTES = 64 * 1024;
export const MAX_SECTION_ELEMENTS = 2 ** 28;
export const MAX_SECTION_RANK = 64;
export const NORMAL_LENGTH_TOLERANCE = 1e-3;

const RULE_IDS = new Set(ruleIds);
const DTYPES = Object.freeze({
  f32: { bytes: 4, ctor: Float32Array },
  f64: { bytes: 8, ctor: Float64Array },
  u32: { bytes: 4, ctor: Uint32Array },
  u16: { bytes: 2, ctor: Uint16Array },
  u8: { bytes: 1, ctor: Uint8Array },
  i32: { bytes: 4, ctor: Int32Array },
});

export class FrameDecodeError extends Error {
  constructor(rule, detail = "") {
    if (!RULE_IDS.has(rule)) throw new Error(`unknown frame rule id: ${rule}`);
    super(detail ? `${rule}: ${detail}` : rule);
    this.name = "FrameDecodeError";
    this.rule = rule;
    this.detail = detail;
  }
}

function fail(rule, detail = "") {
  throw new FrameDecodeError(rule, detail);
}

function align8(value) {
  return Math.ceil(value / 8) * 8;
}

function asBytes(data) {
  if (data instanceof ArrayBuffer) return new Uint8Array(data);
  if (ArrayBuffer.isView(data)) return new Uint8Array(data.buffer, data.byteOffset, data.byteLength);
  throw new TypeError("data must be an ArrayBuffer or an ArrayBuffer view");
}

function isObject(value) {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function asSafeInteger(value) {
  return typeof value === "number" && Number.isSafeInteger(value) ? value : null;
}

function descriptors(header, payloadLength) {
  if (!Array.isArray(header.sections)) fail("sections-invalid", "sections must be a list");
  const result = [];
  const names = new Set();
  let expectedOffset = 0;

  for (let index = 0; index < header.sections.length; index += 1) {
    const raw = header.sections[index];
    if (!isObject(raw)) fail("sections-invalid", `section ${index} must be an object`);
    const { name } = raw;
    if (typeof name !== "string" || name.length === 0) fail("section-name", String(index));
    if (names.has(name)) fail("section-duplicate", name);
    names.add(name);

    const dtype = typeof raw.dtype === "string" ? DTYPES[raw.dtype] : undefined;
    if (!dtype) fail("section-dtype", `${name}: ${String(raw.dtype)}`);
    if (!Array.isArray(raw.shape) || raw.shape.length > MAX_SECTION_RANK) {
      fail("section-shape", name);
    }
    const shape = raw.shape.map(asSafeInteger);
    if (shape.some((dim) => dim === null || dim < 0)) fail("section-shape", name);
    let count = 1;
    for (const dim of shape) {
      if (dim > MAX_SECTION_ELEMENTS) fail("section-elements", name);
      count *= dim;
      if (count > MAX_SECTION_ELEMENTS) fail("section-elements", name);
    }
    const offset = asSafeInteger(raw.byteOffset);
    const length = asSafeInteger(raw.byteLength);
    if (offset === null || length === null || offset < 0 || length < 0) {
      fail("section-range", name);
    }
    if (length !== count * dtype.bytes) fail("section-length", name);
    if (offset > payloadLength || length > payloadLength - offset) fail("section-range", name);
    if (offset % 8 !== 0) fail("section-alignment", name);
    if (offset !== expectedOffset) {
      if (offset < expectedOffset) fail("section-overlap", name);
      fail("section-order", name);
    }
    expectedOffset = align8(offset + length);
    result.push({ name, dtypeName: raw.dtype, dtype, shape, offset, length, count });
  }
  if (payloadLength !== expectedOffset) fail("section-range", "payload has missing or trailing bytes");
  return result;
}

function requireSection(meta, surface, field, missingRule) {
  const name = surface[field];
  if (typeof name !== "string" || name.length === 0 || !meta.has(name)) fail(missingRule, String(name));
  return meta.get(name);
}

function validateFidelity(header) {
  const value = header.fidelity;
  if (!isObject(value)) fail("fidelity-invalid", "preview fidelity must be an object");
  const floatFields = [
    "maxChordErrorMmRequested",
    "maxNormalStepDegRequested",
    "maxNormalStepDegAchieved",
  ];
  const intFields = ["minSilhouetteSegmentsRequested", "minSilhouetteSegmentsAchieved"];
  for (const field of floatFields) {
    if (typeof value[field] !== "number" || !Number.isFinite(value[field]) || value[field] < 0) {
      fail("fidelity-invalid", field);
    }
  }
  for (const field of intFields) {
    if (!Number.isSafeInteger(value[field]) || value[field] < 0) fail("fidelity-invalid", field);
  }
  const complete = value.chordMeasurementComplete;
  const unmeasured = value.unmeasuredChordIntervals;
  if (typeof complete !== "boolean" || !Number.isSafeInteger(unmeasured) || unmeasured < 0) {
    fail("fidelity-invalid", "chord measurement completeness");
  }
  if (complete) {
    if (typeof value.maxChordErrorMmAchieved !== "number" || !Number.isFinite(value.maxChordErrorMmAchieved)
      || value.maxChordErrorMmAchieved < 0 || unmeasured !== 0) fail("fidelity-invalid", "maxChordErrorMmAchieved");
  } else if (value.maxChordErrorMmAchieved !== null || unmeasured < 1) {
    fail("fidelity-invalid", "maxChordErrorMmAchieved");
  }
  if (value.vertexCapLimited !== undefined && typeof value.vertexCapLimited !== "boolean") {
    fail("fidelity-invalid", "vertexCapLimited");
  }
  const missed = !complete
    || (value.maxChordErrorMmAchieved !== null && value.maxChordErrorMmAchieved > value.maxChordErrorMmRequested)
    || value.maxNormalStepDegAchieved > value.maxNormalStepDegRequested
    || value.minSilhouetteSegmentsAchieved < value.minSilhouetteSegmentsRequested;
  if (missed && value.vertexCapLimited !== true) fail("fidelity-cap-unreported");
}

function validatePreview(header, sections, descriptorList) {
  if (!Array.isArray(header.surfaces)) fail("preview-surfaces", "surfaces must be a list");
  validateFidelity(header);
  const meta = new Map(descriptorList.map((item) => [item.name, item]));

  for (let surfaceIndex = 0; surfaceIndex < header.surfaces.length; surfaceIndex += 1) {
    const surface = header.surfaces[surfaceIndex];
    if (!isObject(surface)) fail("surface-invalid", String(surfaceIndex));
    if (typeof surface.role !== "string" || surface.role.length === 0) fail("surface-role", String(surfaceIndex));
    if (surface.shading !== "smooth" && surface.shading !== "flat") fail("surface-shading", surface.role);
    if (!["analytic-parametric", "analytic-spline", "exact-planar"].includes(surface.normalMethod)) {
      fail("surface-normal-method", surface.role);
    }
    if (surface.closedPhi !== undefined && typeof surface.closedPhi !== "boolean") {
      fail("surface-closed-phi", surface.role);
    }

    const positionMeta = requireSection(meta, surface, "positions", "surface-section-missing");
    const indexMeta = requireSection(meta, surface, "indices", "surface-section-missing");
    const normalMeta = requireSection(meta, surface, "normals", "normals-missing");
    if (positionMeta.dtypeName !== "f32" || positionMeta.shape.length !== 2 || positionMeta.shape[1] !== 3) {
      fail("surface-position-format", positionMeta.name);
    }
    if (indexMeta.dtypeName !== "u32" || indexMeta.shape.length !== 1) fail("surface-index-format", indexMeta.name);
    if (normalMeta.dtypeName !== "f32" || normalMeta.shape.length !== 2 || normalMeta.shape[1] !== 3) {
      fail("normals-format", normalMeta.name);
    }
    if (normalMeta.shape[0] !== positionMeta.shape[0]) fail("normals-row-mismatch", surface.role);

    const positions = sections[positionMeta.name];
    const normals = sections[normalMeta.name];
    for (const coordinate of positions) {
      if (!Number.isFinite(coordinate)) fail("positions-nonfinite", positionMeta.name);
    }
    for (let row = 0; row < normalMeta.shape[0]; row += 1) {
      const x = normals[row * 3];
      const y = normals[row * 3 + 1];
      const z = normals[row * 3 + 2];
      if (!Number.isFinite(x) || !Number.isFinite(y) || !Number.isFinite(z)) {
        fail("normals-nonfinite", normalMeta.name);
      }
      if (Math.abs(Math.hypot(x, y, z) - 1) > NORMAL_LENGTH_TOLERANCE) {
        fail("normals-nonunit", normalMeta.name);
      }
    }

    const indices = sections[indexMeta.name];
    const vertexCount = positionMeta.shape[0];
    const check = (start, end) => {
      for (let offset = start; offset < end; offset += 1) {
        if (indices[offset] >= vertexCount) fail("indices-out-of-range", indexMeta.name);
      }
    };
    if (indices.length <= 2048) check(0, indices.length);
    else {
      check(0, 1024);
      check(indices.length - 1024, indices.length);
    }
  }
}

/**
 * Decode a FRAME-SPEC v1.1 message into zero-copy typed-array views.
 * @param {ArrayBuffer|ArrayBufferView} data
 * @param {{maxFrameBytes?: number}} [options]
 * @returns {{header: object, sections: Record<string, TypedArray>}}
 */
export function decodeFrame(data, { maxFrameBytes = DEFAULT_MAX_FRAME_BYTES } = {}) {
  const bytes = asBytes(data);
  if (bytes.byteLength > maxFrameBytes) fail("frame-too-large", `${bytes.byteLength} > ${maxFrameBytes}`);
  if (bytes.byteLength < 4
      || bytes[0] !== 0x57 || bytes[1] !== 0x47 || bytes[2] !== 0x46 || bytes[3] !== 0x31) {
    fail("bad-magic");
  }
  if (bytes.byteLength < 8) fail("header-truncated");
  const dataView = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  const headerLength = dataView.getUint32(4, true);
  if (headerLength > MAX_HEADER_BYTES) fail("header-too-large", String(headerLength));
  const headerEnd = 8 + headerLength;
  if (headerEnd > bytes.byteLength) fail("header-truncated");
  const payloadBase = align8(headerEnd);
  if (payloadBase > bytes.byteLength) fail("header-truncated", "payload alignment padding missing");

  let header;
  try {
    const decoder = new TextDecoder("utf-8", { fatal: true });
    header = JSON.parse(decoder.decode(bytes.subarray(8, headerEnd)));
  } catch (error) {
    fail("header-json", error instanceof Error ? error.message : String(error));
  }
  if (!isObject(header)) fail("header-json", "header must be a JSON object");
  if (asSafeInteger(header.v) !== 1) fail("unsupported-version", String(header.v));
  const descriptorList = descriptors(header, bytes.byteLength - payloadBase);
  const sections = Object.create(null);
  for (const item of descriptorList) {
    const absoluteOffset = bytes.byteOffset + payloadBase + item.offset;
    if (absoluteOffset % item.dtype.bytes !== 0) fail("section-alignment", item.name);
    try {
      sections[item.name] = new item.dtype.ctor(bytes.buffer, absoluteOffset, item.count);
    } catch (error) {
      fail("section-range", error instanceof Error ? error.message : String(error));
    }
  }
  if (header.kind === "preview") validatePreview(header, sections, descriptorList);
  return { header, sections };
}

export const decode = decodeFrame;

File: server/protocol/frame.py
"""FRAME-SPEC v1.1 encoder and zero-copy decoder."""

from __future__ import annotations

import json
import math
import struct
from pathlib import Path
from typing import Any, Mapping, TypeAlias

import numpy as np

Header: TypeAlias = dict[str, Any]

MAGIC = b"WGF1"
SUPPORTED_VERSION = 1
MAX_HEADER_BYTES = 64 * 1024
DEFAULT_MAX_FRAME_BYTES = 32 * 1024 * 1024
MAX_SECTION_ELEMENTS = 1 << 28
MAX_SECTION_RANK = 64
MAX_SAFE_INTEGER = (1 << 53) - 1
NORMAL_LENGTH_TOLERANCE = 1e-3

_DTYPES: dict[str, np.dtype[Any]] = {
    "f32": np.dtype("<f4"),
    "f64": np.dtype("<f8"),
    "u32": np.dtype("<u4"),
    "u16": np.dtype("<u2"),
    "u8": np.dtype("u1"),
    "i32": np.dtype("<i4"),
}
_DTYPE_NAMES = {(dtype.kind, dtype.itemsize): name for name, dtype in _DTYPES.items()}
_RULES_PATH = Path(__file__).resolve().parents[2] / "shared" / "frame-rules.json"
RULE_IDS = frozenset(json.loads(_RULES_PATH.read_text(encoding="utf-8")))


class FrameError(ValueError):
    """A cleanly classified FRAME-SPEC validation failure."""

    def __init__(self, rule: str, detail: str = "") -> None:
        if rule not in RULE_IDS:
            raise RuntimeError(f"unknown frame rule id: {rule}")
        self.rule = rule
        self.detail = detail
        super().__init__(f"{rule}: {detail}" if detail else rule)


def _fail(rule: str, detail: str = "") -> None:
    raise FrameError(rule, detail)


def _align8(value: int) -> int:
    return (value + 7) & ~7


def _safe_integer(value: object) -> int | None:
    """Return the normalized JSON safe integer shared with the JS decoder."""

    if isinstance(value, bool) or not isinstance(value, (int, float)):
        return None
    if isinstance(value, float) and (not math.isfinite(value) or not value.is_integer()):
        return None
    if abs(value) > MAX_SAFE_INTEGER:
        return None
    return int(value)


def _is_int(value: object) -> bool:
    return _safe_integer(value) is not None


def _is_finite_number(value: object) -> bool:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        return False
    try:
        return math.isfinite(value)
    except OverflowError:
        return False


def _json_object(raw: bytes) -> Header:
    try:
        value = json.loads(
            raw.decode("utf-8"),
            parse_constant=lambda value: (_ for _ in ()).throw(ValueError(value)),
        )
    except (UnicodeDecodeError, json.JSONDecodeError, RecursionError, ValueError) as exc:
        _fail("header-json", str(exc))
    if not isinstance(value, dict):
        _fail("header-json", "header must be a JSON object")
    return value


def _section_descriptors(
    header: Header, payload_length: int
) -> list[tuple[str, np.dtype[Any], tuple[int, ...], int, int]]:
    raw_sections = header.get("sections")
    if not isinstance(raw_sections, list):
        _fail("sections-invalid", "sections must be a list")

    result: list[tuple[str, np.dtype[Any], tuple[int, ...], int, int]] = []
    names: set[str] = set()
    ranges: list[tuple[int, int, str]] = []
    expected_offset = 0

    for index, raw in enumerate(raw_sections):
        if not isinstance(raw, dict):
            _fail("sections-invalid", f"section {index} must be an object")
        name = raw.get("name")
        if not isinstance(name, str) or not name:
            _fail("section-name", f"section {index} has no valid name")
        if name in names:
            _fail("section-duplicate", name)
        names.add(name)

        dtype_name = raw.get("dtype")
        if not isinstance(dtype_name, str) or dtype_name not in _DTYPES:
            _fail("section-dtype", f"{name}: {dtype_name!r}")
        dtype = _DTYPES[dtype_name]

        raw_shape = raw.get("shape")
        if not isinstance(raw_shape, list) or len(raw_shape) > MAX_SECTION_RANK:
            _fail("section-shape", name)
        shape_values = [_safe_integer(dim) for dim in raw_shape]
        if any(dim is None or dim < 0 for dim in shape_values):
            _fail("section-shape", name)
        element_count = 1
        for dimension in shape_values:
            assert dimension is not None
            if dimension > MAX_SECTION_ELEMENTS:
                _fail("section-elements", name)
            dim = dimension
            element_count *= dim
            if element_count > MAX_SECTION_ELEMENTS:
                _fail("section-elements", name)
        shape = tuple(dimension for dimension in shape_values if dimension is not None)
        raw["shape"] = list(shape)

        byte_offset = _safe_integer(raw.get("byteOffset"))
        byte_length = _safe_integer(raw.get("byteLength"))
        if byte_offset is None or byte_length is None or byte_offset < 0 or byte_length < 0:
            _fail("section-range", name)
        raw["byteOffset"] = byte_offset
        raw["byteLength"] = byte_length
        if byte_length != element_count * dtype.itemsize:
            _fail("section-length", name)
        if byte_offset > payload_length or byte_length > payload_length - byte_offset:
            _fail("section-range", name)
        if byte_offset % 8 != 0:
            _fail("section-alignment", name)

        if byte_offset != expected_offset:
            if byte_offset < expected_offset:
                _fail("section-overlap", name)
            _fail("section-order", name)
        end = byte_offset + byte_length
        for other_start, other_end, other_name in ranges:
            if byte_offset < other_end and other_start < end:
                _fail("section-overlap", f"{name} overlaps {other_name}")
        ranges.append((byte_offset, end, name))
        expected_offset = _align8(end)
        result.append((name, dtype, shape, byte_offset, byte_length))
    if payload_length != expected_offset:
        _fail("section-range", "payload has missing or trailing bytes")
    return result


def _require_section(
    section_meta: Mapping[str, tuple[np.dtype[Any], tuple[int, ...]]],
    surface: Mapping[str, Any],
    field: str,
    missing_rule: str,
) -> tuple[str, np.dtype[Any], tuple[int, ...]]:
    name = surface.get(field)
    if not isinstance(name, str) or not name:
        _fail(missing_rule, f"surface {field} reference missing")
    try:
        dtype, shape = section_meta[name]
    except KeyError:
        _fail(missing_rule, name)
    return name, dtype, shape


def _validate_fidelity(header: Header) -> None:
    fidelity = header.get("fidelity")
    if not isinstance(fidelity, dict):
        _fail("fidelity-invalid", "preview fidelity must be an object")

    float_fields = (
        "maxChordErrorMmRequested",
        "maxNormalStepDegRequested",
        "maxNormalStepDegAchieved",
    )
    int_fields = ("minSilhouetteSegmentsRequested", "minSilhouetteSegmentsAchieved")
    for field in float_fields:
        value = fidelity.get(field)
        if not _is_finite_number(value) or value < 0:
            _fail("fidelity-invalid", field)
    achieved_chord = fidelity.get("maxChordErrorMmAchieved")
    complete = fidelity.get("chordMeasurementComplete")
    unmeasured = _safe_integer(fidelity.get("unmeasuredChordIntervals"))
    if not isinstance(complete, bool) or unmeasured is None or unmeasured < 0:
        _fail("fidelity-invalid", "chord measurement completeness")
    if complete:
        if not _is_finite_number(achieved_chord) or achieved_chord < 0 or unmeasured != 0:
            _fail("fidelity-invalid", "maxChordErrorMmAchieved")
    elif (
        "maxChordErrorMmAchieved" not in fidelity
        or achieved_chord is not None
        or unmeasured < 1
    ):
        _fail("fidelity-invalid", "maxChordErrorMmAchieved")
    for field in int_fields:
        value = fidelity.get(field)
        if not _is_int(value) or value < 0:
            _fail("fidelity-invalid", field)

    missed = (
        not complete
        or (
            achieved_chord is not None
            and achieved_chord > fidelity["maxChordErrorMmRequested"]
        )
        or fidelity["maxNormalStepDegAchieved"] > fidelity["maxNormalStepDegRequested"]
        or fidelity["minSilhouetteSegmentsAchieved"] < fidelity["minSilhouetteSegmentsRequested"]
    )
    cap_flag = fidelity.get("vertexCapLimited")
    if cap_flag is not None and not isinstance(cap_flag, bool):
        _fail("fidelity-invalid", "vertexCapLimited")
    if missed and cap_flag is not True:
        _fail("fidelity-cap-unreported")


def _validate_preview(header: Header, arrays: Mapping[str, np.ndarray]) -> None:
    surfaces = header.get("surfaces")
    if not isinstance(surfaces, list):
        _fail("preview-surfaces", "surfaces must be a list")
    _validate_fidelity(header)

    section_meta = {name: (array.dtype, array.shape) for name, array in arrays.items()}
    for surface_index, surface in enumerate(surfaces):
        if not isinstance(surface, dict):
            _fail("surface-invalid", str(surface_index))
        role = surface.get("role")
        if not isinstance(role, str) or not role:
            _fail("surface-role", str(surface_index))
        shading = surface.get("shading")
        if not isinstance(shading, str) or shading not in {"smooth", "flat"}:
            _fail("surface-shading", role)
        normal_method = surface.get("normalMethod")
        if not isinstance(normal_method, str) or normal_method not in {
            "analytic-parametric",
            "analytic-spline",
            "exact-planar",
        }:
            _fail("surface-normal-method", role)
        if "closedPhi" in surface and not isinstance(surface["closedPhi"], bool):
            _fail("surface-closed-phi", role)

        position_name, position_dtype, position_shape = _require_section(
            section_meta, surface, "positions", "surface-section-missing"
        )
        index_name, index_dtype, index_shape = _require_section(
            section_meta, surface, "indices", "surface-section-missing"
        )
        normal_name, normal_dtype, normal_shape = _require_section(
            section_meta, surface, "normals", "normals-missing"
        )
        if position_dtype != _DTYPES["f32"] or len(position_shape) != 2 or position_shape[1] != 3:
            _fail("surface-position-format", position_name)
        if index_dtype != _DTYPES["u32"] or len(index_shape) != 1:
            _fail("surface-index-format", index_name)
        if normal_dtype != _DTYPES["f32"] or len(normal_shape) != 2 or normal_shape[1] != 3:
            _fail("normals-format", normal_name)
        if normal_shape[0] != position_shape[0]:
            _fail("normals-row-mismatch", role)

        positions = arrays[position_name]
        normals = arrays[normal_name]
        if not np.isfinite(positions).all():
            _fail("positions-nonfinite", position_name)
        if not np.isfinite(normals).all():
            _fail("normals-nonfinite", normal_name)
        lengths = np.linalg.norm(normals.astype(np.float64, copy=False), axis=1)
        if np.any(np.abs(lengths - 1.0) > NORMAL_LENGTH_TOLERANCE):
            _fail("normals-nonunit", normal_name)

        indices = arrays[index_name]
        if indices.size:
            sample = indices if indices.size <= 2048 else np.concatenate((indices[:1024], indices[-1024:]))
            if np.any(sample >= position_shape[0]):
                _fail("indices-out-of-range", index_name)


def decode(
    data: bytes | bytearray | memoryview, *, max_frame_bytes: int = DEFAULT_MAX_FRAME_BYTES
) -> tuple[Header, dict[str, np.ndarray]]:
    """Decode a frame, returning NumPy views backed by *data*."""

    try:
        view = memoryview(data).cast("B")
    except (TypeError, ValueError) as exc:
        raise TypeError("data must be a contiguous bytes-like object") from exc
    if len(view) > max_frame_bytes:
        _fail("frame-too-large", f"{len(view)} > {max_frame_bytes}")
    if len(view) < 4 or bytes(view[:4]) != MAGIC:
        _fail("bad-magic")
    if len(view) < 8:
        _fail("header-truncated")
    header_length = struct.unpack_from("<I", view, 4)[0]
    if header_length > MAX_HEADER_BYTES:
        _fail("header-too-large", str(header_length))
    header_end = 8 + header_length
    if header_end > len(view):
        _fail("header-truncated")
    payload_base = _align8(header_end)
    if payload_base > len(view):
        _fail("header-truncated", "payload alignment padding missing")

    header = _json_object(bytes(view[8:header_end]))
    version = _safe_integer(header.get("v"))
    if version != SUPPORTED_VERSION:
        _fail("unsupported-version", repr(version))
    header["v"] = version
    descriptors = _section_descriptors(header, len(view) - payload_base)

    arrays: dict[str, np.ndarray] = {}
    for name, dtype, shape, byte_offset, _byte_length in descriptors:
        try:
            arrays[name] = np.ndarray(
                shape, dtype=dtype, buffer=view, offset=payload_base + byte_offset
            )
        except (TypeError, ValueError, OverflowError) as exc:
            _fail("section-shape", f"{name}: {exc}")
    if header.get("kind") == "preview":
        _validate_preview(header, arrays)
    return header, arrays


def encode(header_fields: Mapping[str, Any], arrays: Mapping[str, np.ndarray]) -> bytes:
    """Encode a deterministic FRAME-SPEC v1.1 frame."""

    if not isinstance(header_fields, Mapping) or not isinstance(arrays, Mapping):
        raise TypeError("header_fields and arrays must be mappings")
    header: Header = dict(header_fields)
    if "sections" in header:
        _fail("sections-invalid", "encoder owns the sections field")
    header.setdefault("v", SUPPORTED_VERSION)

    payload = bytearray()
    sections: list[dict[str, Any]] = []
    for name, value in arrays.items():
        if not isinstance(name, str) or not name:
            _fail("section-name")
        array = np.asarray(value)
        dtype_name = _DTYPE_NAMES.get((array.dtype.kind, array.dtype.itemsize))
        if dtype_name is None:
            _fail("section-dtype", f"{name}: {array.dtype}")
        dtype = _DTYPES[dtype_name]
        element_count = int(array.size)
        if element_count > MAX_SECTION_ELEMENTS:
            _fail("section-elements", name)
        semantic_name = name.lower()
        if (semantic_name.endswith("positions") or semantic_name.endswith("normals")) and not np.isfinite(array).all():
            _fail("normals-nonfinite" if semantic_name.endswith("normals") else "positions-nonfinite", name)
        while len(payload) % 8:
            payload.append(0)
        raw = array.astype(dtype, copy=False).tobytes(order="C")
        sections.append(
            {
                "name": name,
                "dtype": dtype_name,
                "shape": list(array.shape),
                "byteOffset": len(payload),
                "byteLength": len(raw),
            }
        )
        payload.extend(raw)
    while len(payload) % 8:
        payload.append(0)
    header["sections"] = sections
    try:
        header_bytes = json.dumps(
            header, ensure_ascii=False, allow_nan=False, separators=(",", ":")
        ).encode("utf-8")
    except (RecursionError, TypeError, ValueError) as exc:
        _fail("header-json", str(exc))
    if len(header_bytes) > MAX_HEADER_BYTES:
        _fail("header-too-large", str(len(header_bytes)))
    prefix = MAGIC + struct.pack("<I", len(header_bytes)) + header_bytes
    frame = prefix + bytes(_align8(len(prefix)) - len(prefix)) + payload
    if len(frame) > DEFAULT_MAX_FRAME_BYTES:
        _fail("frame-too-large", str(len(frame)))
    decode(frame)
    return frame
